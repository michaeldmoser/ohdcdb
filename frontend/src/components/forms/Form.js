import React from 'react';
import { Formik } from 'formik';
import { Form as BootstrapForm } from 'react-bootstrap';

export function Form({ children, ...rest }) {
    return (
        <Formik {...rest}>
            {(formik) => {
                return (
                    <BootstrapForm noValidate onSubmit={formik.handleSubmit}>
                        {children(formik)}
                    </BootstrapForm>
                );
            }}
        </Formik>
    );
}
