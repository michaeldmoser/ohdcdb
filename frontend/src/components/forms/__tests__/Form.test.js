import { render, userEvent, screen, act } from 'testing/library';
import { Form } from '../Form';
import * as Yup from 'yup';
import { TextInput, SubmitButton } from '..';

it('should render a form', () => {
    const form = render(
        <Form
            initialValues={{
                title: '',
                description: '',
            }}
            onSubmit={(values) => {}}
        >
            {() => (
                <>
                    <TextInput name='title' />
                    <TextInput name='description' />
                    <SubmitButton>Save</SubmitButton>
                </>
            )}
        </Form>
    );

    expect(form).toMatchSnapshot();
});

it('should complain about field being required.', async () => {
    const form = render(
        <Form
            initialValues={{
                title: '',
                description: '',
            }}
            validationSchema={Yup.object({
                title: Yup.string().required('Required'),
            })}
            onSubmit={(values) => {}}
        >
            {() => (
                <>
                    <TextInput name='title' />
                    <TextInput name='description' />
                    <SubmitButton>Save</SubmitButton>
                </>
            )}
        </Form>
    );

    await act(async () => {
        userEvent.click(screen.getByLabelText(/title/i));
        userEvent.click(screen.getByLabelText(/description/i));
    });

    expect(form).toMatchSnapshot();
});

it('should submit the form', async () => {
    let actualValues = null;
    const expectedValues = {
        title: 'The title',
        description: 'The description',
    };

    render(
        <Form
            initialValues={{
                title: '',
                description: '',
            }}
            onSubmit={(values) => {
                actualValues = values;
            }}
        >
            {() => (
                <>
                    <TextInput name='title' />
                    <TextInput name='description' />
                    <SubmitButton>Save</SubmitButton>
                </>
            )}
        </Form>
    );

    await act(async () => {
        userEvent.type(screen.getByLabelText(/title/i), expectedValues.title);
        userEvent.type(
            screen.getByLabelText(/description/i),
            expectedValues.description
        );
        userEvent.click(screen.getByRole('button', { name: /save/i }));
    });

    expect(actualValues).toEqual(expectedValues);
});

it('should not submit if not filled out correctly', async () => {
    let submitted = false;

    render(
        <Form
            initialValues={{
                title: '',
                description: '',
            }}
            validationSchema={Yup.object({
                title: Yup.string().required('Required'),
            })}
            onSubmit={(values) => {
                submitted = true;
            }}
        >
            {() => (
                <>
                    <TextInput name='title' />
                    <TextInput name='description' />
                    <SubmitButton>Save</SubmitButton>
                </>
            )}
        </Form>
    );

    await act(async () => {
        userEvent.click(screen.getByRole('button', { name: /save/i }));
    });

    expect(submitted).toBeFalse();
});
