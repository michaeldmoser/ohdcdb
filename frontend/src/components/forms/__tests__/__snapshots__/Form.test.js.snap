// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should complain about field being required. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class=""
        novalidate=""
      >
        <div
          class="mb-3 form-floating col"
        >
          <input
            class="form-control is-invalid"
            feedback="Required"
            id="title"
            name="title"
            placeholder="Title..."
            type="text"
            value=""
          />
          <div
            class="invalid-tooltip"
          >
            Required
          </div>
          <label
            for="title"
          >
            Title
          </label>
        </div>
        <div
          class="mb-3 form-floating col"
        >
          <input
            class="form-control"
            id="description"
            name="description"
            placeholder="Description..."
            type="text"
            value=""
          />
          <div
            class="invalid-tooltip"
          />
          <label
            for="description"
          >
            Description
          </label>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class=""
      novalidate=""
    >
      <div
        class="mb-3 form-floating col"
      >
        <input
          class="form-control is-invalid"
          feedback="Required"
          id="title"
          name="title"
          placeholder="Title..."
          type="text"
          value=""
        />
        <div
          class="invalid-tooltip"
        >
          Required
        </div>
        <label
          for="title"
        >
          Title
        </label>
      </div>
      <div
        class="mb-3 form-floating col"
      >
        <input
          class="form-control"
          id="description"
          name="description"
          placeholder="Description..."
          type="text"
          value=""
        />
        <div
          class="invalid-tooltip"
        />
        <label
          for="description"
        >
          Description
        </label>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should render a form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class=""
        novalidate=""
      >
        <div
          class="mb-3 form-floating col"
        >
          <input
            class="form-control"
            id="title"
            name="title"
            placeholder="Title..."
            type="text"
            value=""
          />
          <div
            class="invalid-tooltip"
          />
          <label
            for="title"
          >
            Title
          </label>
        </div>
        <div
          class="mb-3 form-floating col"
        >
          <input
            class="form-control"
            id="description"
            name="description"
            placeholder="Description..."
            type="text"
            value=""
          />
          <div
            class="invalid-tooltip"
          />
          <label
            for="description"
          >
            Description
          </label>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class=""
      novalidate=""
    >
      <div
        class="mb-3 form-floating col"
      >
        <input
          class="form-control"
          id="title"
          name="title"
          placeholder="Title..."
          type="text"
          value=""
        />
        <div
          class="invalid-tooltip"
        />
        <label
          for="title"
        >
          Title
        </label>
      </div>
      <div
        class="mb-3 form-floating col"
      >
        <input
          class="form-control"
          id="description"
          name="description"
          placeholder="Description..."
          type="text"
          value=""
        />
        <div
          class="invalid-tooltip"
        />
        <label
          for="description"
        >
          Description
        </label>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
